toothpick-runtime/src/main/java/toothpick/ToothPick.java

```diff
@@ -68,26 +76,25 @@ public static Scope openScope(Object name) {
    *
    * @param name the name of the scope to close.
    */
-  public static void closeScope(Object name) {
-    synchronized (ToothPick.class) {
-      Scope scope = MAP_KEY_TO_SCOPE.get(name);
-      if (scope != null) {
-        Scope parentScope = scope.getParentScope();
-        if (parentScope != null) {
-          parentScope.removeChild(scope);
-        }
-        removeScopeAndChildrenFromMap(scope);
+
+  public static synchronized void closeScope(Object name) {
+    //we remove the scope first, so that other threads don't see it, and see the next snapshot of the tree
+    Scope scope = MAP_KEY_TO_SCOPE.remove(name);
+    if (scope != null) {
+      Scope parentScope = scope.getParentScope();
+      if (parentScope != null) {
+        parentScope.removeChild(scope);
       }
+      removeScopeAndChildrenFromMap(scope);
     }
   }
 
   /**
    * Clears all scopes. Useful for testing and not getting any leak...
    */
+
   public static void reset() {
```

Comment:
it's needed for tests. That was one of the reasons of the latest leaks in RG during testing.